zebra-network/src/peer/client.rs

```diff
@@ -253,31 +261,53 @@ impl<T: std::fmt::Debug> Drop for MustUseOneshotSender<T> {
 impl Client {
     /// Check if this connection's heartbeat task has exited.
     fn check_heartbeat(&mut self, cx: &mut Context<'_>) -> Result<(), SharedPeerError> {
-        if let Poll::Ready(()) = self
+        let is_canceled = self
             .shutdown_tx
             .as_mut()
             .expect("only taken on drop")
             .poll_canceled(cx)
-        {
-            // Make sure there is an error in the slot
-            let heartbeat_error: SharedPeerError = PeerError::HeartbeatTaskExited.into();
-            let original_error = self.error_slot.try_update_error(heartbeat_error.clone());
-            debug!(
-                ?original_error,
-                latest_error = ?heartbeat_error,
-                "client heartbeat task exited"
-            );
+            .is_ready();
 
-            if let Err(AlreadyErrored { original_error }) = original_error {
-                Err(original_error)
-            } else {
-                Err(heartbeat_error)
-            }
+        let has_stopped = self.heartbeat_task.poll_unpin(cx).is_ready();
+
+        if is_canceled || has_stopped {
+            self.set_task_exited_error("heartbeat", PeerError::HeartbeatTaskExited)
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Check if the connection's task has exited.
+    fn check_connection(&mut self, context: &mut Context<'_>) -> Result<(), SharedPeerError> {
+        if self.connection_task.poll_unpin(context).is_ready() {
+            self.set_task_exited_error("heartbeat", PeerError::HeartbeatTaskExited)
```

Comment:
Is this a copy-paste error?

```suggestion
            self.set_task_exited_error("connection", PeerError::ConnectionTaskExited)
```

Reply:
Yes. Fixed, thanks :+1: